pkg/executor/executor.go

```diff
@@ -1246,8 +1250,16 @@ func newLockCtx(sctx sessionctx.Context, lockWaitTime int64, numKeys int) (*tikv
 		rec := deadlockhistory.ErrDeadlockToDeadlockRecord(deadlock)
 		deadlockhistory.GlobalDeadlockHistory.Push(rec)
 	}
+	lockCtx.OnMemChange = func(capacity int) uint64 {
+		bucketMemoryUsage := hack.EstimateBucketMemoryUsageWithKVSize(lockCtx.GetValuesKSize(), lockCtx.GetValuesVSize())
+		mapSize := set.EstimateMapSize(capacity, bucketMemoryUsage)
+		return mapSize
+	}
 	if lockCtx.ForUpdateTS > 0 && seVars.AssertionLevel != variable.AssertionLevelOff {
 		lockCtx.InitCheckExistence(numKeys)
+		if lockCtx.OnMemChange != nil {
```

Comment:
OnMemChange is defined in this function and only used here. So it seems to me that it is redundant to have a OnMemChange field in LockCtx. 

So you are only recording the initial usage of the struct. That leads to a question: should we dynamically track the structure, or is the initial usage is enough? Anyway it's a good start and better than tracking nothing.